MSTSC crashes when resolution is changed by maximizing on a different monitor

I have two monitors with 2560x1440 and 1920x1080 resolution.

and setting mstsc to use one monitor for full screen mode.

when I drag mstsc from a monitor to another, error occurred as below.
![image](https://user-images.githubusercontent.com/3352585/123057364-63a6ce80-d43a-11eb-8577-0656247746ac.png)
It means internal error.

and will got log 
```

==> /home/.xorgxrdp.12.log <==
[1135171.145] rdpClientConProcessMsgClientInput: invalidate x 0 y 0 cx 2560 cy 1440
[1135171.145] rdpClientConProcessMsgVersion: version 0 0 0 1
[1135171.145] rdpClientConProcessScreenSizeMsg: set width 2560 height 1440 bpp 32
[1135171.146] rdpClientConProcessScreenSizeMsg: shmemid 254443553 shmemptr 0x7f5f87b3a000
[1135171.146] rdpRRScreenSetSize: width 2560 height 1440 mmWidth 677 mmHeight 381
[1135171.151] rdpRRGetInfo:
[1135171.151]   screen resized to 2560x1440
[1135171.157] rdpClientConProcessScreenSizeMsg: RRScreenSizeSet ok=[1]
[1135171.157] rdpClientConProcessMsgClientInput: invalidate x 0 y 0 cx 2560 cy 1440
[1135171.157] rdpClientConProcessMsgClientInput: invalidate x 0 y 0 cx 2560 cy 1440
[1135171.158] rdpClientConProcessMsgVersion: version 0 0 0 1
[1135171.158] rdpClientConProcessScreenSizeMsg: set width 2560 height 1440 bpp 32
[1135171.158] rdpClientConProcessScreenSizeMsg: shmemid 254476321 shmemptr 0x7f5f87b3a000
[1135171.158] rdpClientConProcessMsgClientInput: invalidate x 0 y 0 cx 2560 cy 1440
[1135171.158] rdpClientConProcessMsgClientInfo:
[1135171.158]   got client info bytes 7064
[1135171.158]   jpeg support 0
[1135171.158]   offscreen support 1
[1135171.158]   offscreen size 10485760
[1135171.158]   offscreen entries 100
[1135171.158] rdpClientConProcessMsgClientInfo: got RFX capture
[1135171.158]   cap_width 2560 cap_height 1472
[1135171.158] rdpClientConProcessMsgClientInfo: shmemid 254509089 shmemptr 0x7f5f87aea000 bytes 15073280
[1135171.158]   client can not do offscreen to offscreen blits
[1135171.158]   client can do new(color) cursor
[1135171.158]   client can not do multimon
[1135171.158] rdpRRSetRdpOutputs: numCrtcs 1 numOutputs 1 monitorCount 0
[1135171.158] rdpRRSetRdpOutputs: update output 0 left 0 top 0 width 2560 height 1440
[1135171.158] rdpRRUpdateOutput:
[1135171.159] rdpLoadLayout: keylayout 0x00000804 variant  display 12
[1135171.159] rdpkeybChangeKeyboardControl:
[1135171.159] rdpkeybChangeKeyboardControl: autoRepeat on
[1135171.159] rdpkeybChangeKeyboardControl:
[1135171.160] rdpkeybChangeKeyboardControl: autoRepeat on
[1135171.160] rdpClientConProcessMsgClientInfo:
[1135171.160]   got client info bytes 7064
[1135171.160]   jpeg support 0
[1135171.160]   offscreen support 1
[1135171.160]   offscreen size 10485760
[1135171.160]   offscreen entries 100
[1135171.160] rdpClientConProcessMsgClientInfo: got RFX capture
[1135171.160]   cap_width 2560 cap_height 1472
[1135171.160] rdpClientConProcessMsgClientInfo: shmemid 254541857 shmemptr 0x7f5f87aea000 bytes 15073280
[1135171.160]   client can not do offscreen to offscreen blits
[1135171.161]   client can do new(color) cursor
[1135171.161]   client can not do multimon
[1135171.161] rdpRRSetRdpOutputs: numCrtcs 1 numOutputs 1 monitorCount 0
[1135171.161] rdpRRSetRdpOutputs: update output 0 left 0 top 0 width 2560 height 1440
[1135171.161] rdpRRUpdateOutput:
[1135171.161] rdpLoadLayout: keylayout 0x00000804 variant  display 12
[1135171.161] rdpkeybChangeKeyboardControl:
[1135171.161] rdpkeybChangeKeyboardControl: autoRepeat on
[1135171.161] rdpkeybChangeKeyboardControl:
[1135171.161] rdpkeybChangeKeyboardControl: autoRepeat on
[1135171.240] rdpRRGetInfo:
[1135171.259] rdpInDeferredRepeatCallback:
[1135171.259] rdpkeybChangeKeyboardControl:
[1135171.259] rdpkeybChangeKeyboardControl: autoRepeat off
[1135171.260] rdpRRGetInfo:
[1135171.260] rdpInDeferredRepeatCallback:
[1135171.260] rdpkeybChangeKeyboardControl:
[1135171.260] rdpkeybChangeKeyboardControl: autoRepeat off
[1135171.268] rdpInDeferredRepeatCallback:
[1135171.268] rdpkeybChangeKeyboardControl:
[1135171.268] rdpkeybChangeKeyboardControl: autoRepeat off
[1135171.268] rdpInDeferredRepeatCallback:
[1135171.268] rdpkeybChangeKeyboardControl:
[1135171.268] rdpkeybChangeKeyboardControl: autoRepeat off
[1135171.273] rdpRRGetInfo:

==> /var/log/xrdp.log <==
[20210623-15:44:38] [INFO ] xrdp_caps_process_pointer: client supports new(color) cursor
[20210623-15:44:38] [INFO ] xrdp_process_offscreen_bmpcache: support level 1 cache size 10485760 MB cache entries 100
[20210623-15:44:38] [INFO ] xrdp_caps_process_codecs: nscodec, codec id 1, properties len 3
[20210623-15:44:38] [WARN ] xrdp_caps_process_codecs: unknown codec id 5
[20210623-15:44:38] [INFO ] xrdp_caps_process_codecs: RemoteFX, codec id 3, properties len 49

==> /home/.xorgxrdp.12.log <==
[1135171.458] KbdSync: toggling num lock
[1135171.458] rdpkeybChangeKeyboardControl:
[1135171.458] rdpkeybChangeKeyboardControl: autoRepeat off
[1135171.458] rdpkeybChangeKeyboardControl:
[1135171.458] rdpkeybChangeKeyboardControl: autoRepeat off
[1135171.486] rdpClientConProcessMsgClientInput: invalidate x 0 y 0 cx 2559 cy 1439

==> /var/log/xrdp.log <==
[20210623-15:44:38] [ERROR] SSL_read: I/O error
[20210623-15:44:38] [ERROR] xrdp_iso_send: trans_write_copy_s failed
[20210623-15:44:38] [ERROR] SSL_shutdown: I/O error
[20210623-15:44:38] [ERROR] Sending [ITU T.125] DisconnectProviderUltimatum failed

==> /home/.xorgxrdp.12.log <==
[1135171.521] rdpClientConRecv: g_sck_recv failed(returned -1)
[1135171.521] rdpClientConRecvMsg: error
[1135171.521] rdpClientConCheck: rdpClientConGotData failed
[1135171.525] rdpClientConDisconnect:
[1135171.525] rdpRemoveClientConFromDev: removing clientCon 0x55644bb73110
```

These errors from xrdp mean the other client has gone away unexpectedly:-

```
[20210623-15:44:38] [ERROR] SSL_read: I/O error
[20210623-15:44:38] [ERROR] xrdp_iso_send: trans_write_copy_s failed
[20210623-15:44:38] [ERROR] SSL_shutdown: I/O error
[20210623-15:44:38] [ERROR] Sending [ITU T.125] DisconnectProviderUltimatum failed
```

So something is happening which is upsetting `mstsc.exe`, but it's not clear what.

I've got a twin-screen Windows 10 VM set up as you have above, and I'm unable to reproduce your crash. Windows version here is 21H1 (OS build 19043.1052), and I'm using a development build of xrdp.

What platforms are you using? Are you using a .ini file to start the connection?

I've reproduced this by using a LAN setting and 24bpp colour.

The debug logs are required to get some idea of what's going on. Here's what I've got so far from xrdp:-

<details>
<summary>xrdp.log</summary>
<pre>
[20211119-11:57:10] [DEBUG] [dynamic_monitor_data(xrdp_mm.c:1080)] dynamic_monitor_data: msg_type 2 msg_length 56
[20211119-11:57:10] [DEBUG] [dynamic_monitor_data(xrdp_mm.c:1092)]   MonitorLayoutSize 40 NumMonitor 1
[20211119-11:57:10] [DEBUG] [dynamic_monitor_data(xrdp_mm.c:1107)]     Flags 0x00000001 Left 0 Top 0 Width 1920 Height 1200 PhysicalWidth 637 PhysicalHeight 421 Orientation 0 DesktopScaleFactor 100 DeviceScaleFactor 100
[20211119-11:57:10] [DEBUG] [libxrdp_send_pointer(libxrdp.c:719)] sending cursor
[20211119-11:57:10] [DEBUG] [libxrdp_send_pointer(libxrdp.c:747)] libxrdp_send_pointer: fastpath
[20211119-11:57:10] [DEBUG] [xrdp_rdp_send_fastpath(xrdp_rdp.c:785)] xrdp_rdp_send_fastpath: no_comp_len 3221, fragmentation 0
[20211119-11:57:10] [DEBUG] [xrdp_sec_send_fastpath(xrdp_sec.c:1921)] xrdp_sec_send_fastpath: no crypt
[20211119-11:57:10] [DEBUG] [libxrdp_send_pointer(libxrdp.c:719)] sending cursor
[20211119-11:57:10] [DEBUG] [libxrdp_send_pointer(libxrdp.c:747)] libxrdp_send_pointer: fastpath
[20211119-11:57:10] [DEBUG] [xrdp_rdp_send_fastpath(xrdp_rdp.c:785)] xrdp_rdp_send_fastpath: no_comp_len 3221, fragmentation 0
[20211119-11:57:10] [DEBUG] [xrdp_sec_send_fastpath(xrdp_sec.c:1921)] xrdp_sec_send_fastpath: no crypt
[20211119-11:57:10] [DEBUG] [xrdp_painter_create(xrdp_painter.c:140)] xrdp_painter_create:
[20211119-11:57:10] [DEBUG] [xrdp_painter_begin_update(xrdp_painter.c:243)] xrdp_painter_begin_update:
[20211119-11:57:10] [DEBUG] [xrdp_orders_init(xrdp_orders.c:110)] xrdp_orders_init: fastpath
[20211119-11:57:10] [DEBUG] [wm_painter_set_target(xrdp_painter.c:193)] wm_painter_set_target:
[20211119-11:57:10] [DEBUG] [xrdp_painter_end_update(xrdp_painter.c:265)] xrdp_painter_end_update:
[20211119-11:57:10] [DEBUG] [xrdp_painter_delete(xrdp_painter.c:171)] xrdp_painter_delete:
[20211119-11:57:10] [DEBUG] [send_server_monitor_resize(xup.c:1307)] send_server_monitor_resize: sent resize message with following properties to xorgxrdp backend width=1920, height=1200, bpp=24, return value=0
[20211119-11:57:10] [DEBUG] [send_server_monitor_full_invalidate(xup.c:1334)] send_server_monitor_full_invalidate: sent invalidate message with following properties to xorgxrdp backend width=1920, height=1200, return value=0
[20211119-11:57:10] [DEBUG] [xrdp_rdp_recv(xrdp_rdp.c:519)] xrdp_rdp_recv: out code 0 (skip data) data processed by channel id 1007
[20211119-11:57:10] [DEBUG] [lib_mod_process_orders(xup.c:1386)] lib_mod_process_orders: type 1
[20211119-11:57:11] [DEBUG] [xrdp_painter_create(xrdp_painter.c:140)] xrdp_painter_create:
[20211119-11:57:11] [DEBUG] [xrdp_painter_begin_update(xrdp_painter.c:243)] xrdp_painter_begin_update:
[20211119-11:57:11] [DEBUG] [xrdp_orders_init(xrdp_orders.c:110)] xrdp_orders_init: fastpath
[20211119-11:57:11] [DEBUG] [wm_painter_set_target(xrdp_painter.c:193)] wm_painter_set_target:
[20211119-11:57:11] [DEBUG] [lib_mod_process_orders(xup.c:1386)] lib_mod_process_orders: type 61
[20211119-11:57:11] [DEBUG] [server_paint_rects(xrdp_mm.c:3192)] server_paint_rects: 0x55ae03f16af0
[20211119-11:57:11] [DEBUG] [lib_mod_process_orders(xup.c:1386)] lib_mod_process_orders: type 2
[20211119-11:57:11] [DEBUG] [xrdp_painter_end_update(xrdp_painter.c:265)] xrdp_painter_end_update:
[20211119-11:57:11] [DEBUG] [process_enc_rfx(xrdp_encoder.c:329)] process_enc_rfx:
[20211119-11:57:11] [DEBUG] [process_enc_rfx(xrdp_encoder.c:330)] process_enc_rfx: num_crects 3 num_drects 1
[20211119-11:57:11] [DEBUG] [xrdp_painter_delete(xrdp_painter.c:171)] xrdp_painter_delete:
[20211119-11:57:11] [DEBUG] [process_enc_rfx(xrdp_encoder.c:393)] process_enc_rfx: rfxcodec_encode rv 0
[20211119-11:57:11] [DEBUG] [dynamic_monitor_data(xrdp_mm.c:1080)] dynamic_monitor_data: msg_type 2 msg_length 56
[20211119-11:57:11] [DEBUG] [dynamic_monitor_data(xrdp_mm.c:1092)]   MonitorLayoutSize 40 NumMonitor 1
[20211119-11:57:11] [DEBUG] [dynamic_monitor_data(xrdp_mm.c:1107)]     Flags 0x00000001 Left 0 Top 0 Width 1920 Height 1200 PhysicalWidth 637 PhysicalHeight 421 Orientation 0 DesktopScaleFactor 100 DeviceScaleFactor 100
[20211119-11:57:11] [DEBUG] [libxrdp_send_pointer(libxrdp.c:719)] sending cursor
[20211119-11:57:11] [DEBUG] [libxrdp_send_pointer(libxrdp.c:747)] libxrdp_send_pointer: fastpath
[20211119-11:57:11] [DEBUG] [xrdp_rdp_send_fastpath(xrdp_rdp.c:785)] xrdp_rdp_send_fastpath: no_comp_len 3221, fragmentation 0
[20211119-11:57:11] [DEBUG] [xrdp_sec_send_fastpath(xrdp_sec.c:1921)] xrdp_sec_send_fastpath: no crypt
[20211119-11:57:11] [DEBUG] [libxrdp_send_pointer(libxrdp.c:719)] sending cursor
[20211119-11:57:11] [DEBUG] [libxrdp_send_pointer(libxrdp.c:747)] libxrdp_send_pointer: fastpath
[20211119-11:57:11] [DEBUG] [xrdp_rdp_send_fastpath(xrdp_rdp.c:785)] xrdp_rdp_send_fastpath: no_comp_len 3221, fragmentation 0
[20211119-11:57:11] [DEBUG] [xrdp_sec_send_fastpath(xrdp_sec.c:1921)] xrdp_sec_send_fastpath: no crypt
[20211119-11:57:11] [DEBUG] [xrdp_painter_create(xrdp_painter.c:140)] xrdp_painter_create:
[20211119-11:57:11] [DEBUG] [xrdp_painter_begin_update(xrdp_painter.c:243)] xrdp_painter_begin_update:
[20211119-11:57:11] [DEBUG] [xrdp_orders_init(xrdp_orders.c:110)] xrdp_orders_init: fastpath
[20211119-11:57:11] [DEBUG] [wm_painter_set_target(xrdp_painter.c:193)] wm_painter_set_target:
[20211119-11:57:12] [DEBUG] [xrdp_painter_end_update(xrdp_painter.c:265)] xrdp_painter_end_update:
[20211119-11:57:12] [DEBUG] [xrdp_painter_delete(xrdp_painter.c:171)] xrdp_painter_delete:
[20211119-11:57:12] [DEBUG] [send_server_monitor_resize(xup.c:1307)] send_server_monitor_resize: sent resize message with following properties to xorgxrdp backend width=1920, height=1200, bpp=24, return value=0
[20211119-11:57:12] [DEBUG] [send_server_monitor_full_invalidate(xup.c:1334)] send_server_monitor_full_invalidate: sent invalidate message with following properties to xorgxrdp backend width=1920, height=1200, return value=0
[20211119-11:57:12] [DEBUG] [xrdp_rdp_recv(xrdp_rdp.c:519)] xrdp_rdp_recv: out code 0 (skip data) data processed by channel id 1007
[20211119-11:57:12] [DEBUG] [xrdp_mm_process_enc_done(xrdp_mm.c:2815)] xrdp_mm_process_enc_done: message back bytes 2914
[20211119-11:57:12] [DEBUG] [libxrdp_fastpath_send_frame_marker(libxrdp.c:1723)] libxrdp_fastpath_send_frame_marker:
[20211119-11:57:12] [DEBUG] [xrdp_rdp_send_fastpath(xrdp_rdp.c:785)] xrdp_rdp_send_fastpath: no_comp_len 12, fragmentation 0
[20211119-11:57:12] [DEBUG] [xrdp_sec_send_fastpath(xrdp_sec.c:1921)] xrdp_sec_send_fastpath: no crypt
[20211119-11:57:12] [DEBUG] [libxrdp_fastpath_send_surface(libxrdp.c:1641)] libxrdp_fastpath_send_surface:
[20211119-11:57:12] [DEBUG] [xrdp_rdp_send_fastpath(xrdp_rdp.c:785)] xrdp_rdp_send_fastpath: no_comp_len 2940, fragmentation 0
[20211119-11:57:12] [DEBUG] [compress_rdp_5(xrdp_mppc_enc.c:972)] Compression algorithim produced a compressed buffer which is larger than the uncompressed buffer. compression ratio 0.979646, flags 0x1
[20211119-11:57:12] [DEBUG] [xrdp_rdp_send_fastpath(xrdp_rdp.c:810)] compress_rdp failed, sending uncompressed data. type 2, flags 1
[20211119-11:57:12] [DEBUG] [xrdp_sec_send_fastpath(xrdp_sec.c:1921)] xrdp_sec_send_fastpath: no crypt
[20211119-11:57:12] [DEBUG] [libxrdp_fastpath_send_frame_marker(libxrdp.c:1723)] libxrdp_fastpath_send_frame_marker:
[20211119-11:57:12] [DEBUG] [xrdp_rdp_send_fastpath(xrdp_rdp.c:785)] xrdp_rdp_send_fastpath: no_comp_len 12, fragmentation 0
[20211119-11:57:12] [DEBUG] [xrdp_sec_send_fastpath(xrdp_sec.c:1921)] xrdp_sec_send_fastpath: no crypt
<b>[20211119-11:57:12] [ERROR] [ssl_tls_log_error(ssl_calls.c:655)] SSL_write: I/O error</b>
[20211119-11:57:12] [ERROR] [xrdp_sec_send_fastpath(xrdp_sec.c:1936)] xrdp_sec_send_fastpath: xrdp_fastpath_send failed
[20211119-11:57:12] [ERROR] [xrdp_rdp_send_fastpath(xrdp_rdp.c:834)] xrdp_rdp_send_fastpath: xrdp_sec_send_fastpath failed
[20211119-11:57:13] [ERROR] [libxrdp_fastpath_send_frame_marker(libxrdp.c:1749)] libxrdp_fastpath_send_frame_marker: xrdp_rdp_send_fastpath failed
[20211119-11:57:13] [DEBUG] [xrdp_mm_process_enc_done(xrdp_mm.c:2838)] xrdp_mm_process_enc_done: last set
[20211119-11:57:13] [DEBUG] [xrdp_mm_update_module_frame_ack(xrdp_mm.c:2785)] xrdp_mm_update_module_ack: frame_id_server 10
[20211119-11:57:13] [DEBUG] [lib_mod_process_orders(xup.c:1386)] lib_mod_process_orders: type 1
[20211119-11:57:13] [DEBUG] [xrdp_painter_create(xrdp_painter.c:140)] xrdp_painter_create:
[20211119-11:57:13] [DEBUG] [xrdp_painter_begin_update(xrdp_painter.c:243)] xrdp_painter_begin_update:
[20211119-11:57:13] [DEBUG] [xrdp_orders_init(xrdp_orders.c:110)] xrdp_orders_init: fastpath
[20211119-11:57:13] [DEBUG] [wm_painter_set_target(xrdp_painter.c:193)] wm_painter_set_target:
[20211119-11:57:13] [DEBUG] [lib_mod_process_orders(xup.c:1386)] lib_mod_process_orders: type 51
[20211119-11:57:13] [DEBUG] [libxrdp_send_pointer(libxrdp.c:719)] sending cursor
[20211119-11:57:13] [DEBUG] [libxrdp_send_pointer(libxrdp.c:747)] libxrdp_send_pointer: fastpath
[20211119-11:57:13] [DEBUG] [xrdp_rdp_send_fastpath(xrdp_rdp.c:785)] xrdp_rdp_send_fastpath: no_comp_len 4245, fragmentation 0
[20211119-11:57:13] [DEBUG] [xrdp_sec_send_fastpath(xrdp_sec.c:1921)] xrdp_sec_send_fastpath: no crypt
[20211119-11:57:13] [ERROR] [xrdp_sec_send_fastpath(xrdp_sec.c:1936)] xrdp_sec_send_fastpath: xrdp_fastpath_send failed
</pre>
</details>

All I can say at the moment is something seems to be going wrong in `xrdp_rdp_send_fastpath()` following the resize. I've not been able to get any further yet.

--- a/xrdp/xrdp_mm.c
+++ b/xrdp/xrdp_mm.c
@@ -1130,6 +1130,17 @@ process_dynamic_monitor_description(struct xrdp_wm *wm,
         }
     }
 
+    // Need to recreate the encoder for connections that use it.
+    if (mm->encoder != NULL)
+    {
+        xrdp_encoder_delete(mm->encoder);
+        mm->encoder = NULL;
+    }
+    if (mm->encoder == NULL)
+    {
+        mm->encoder = xrdp_encoder_create(mm);
+    }
+
     wm->client_info->display_sizes.monitorCount = description->monitorCount;
     wm->client_info->display_sizes.session_width = description->session_width;
     wm->client_info->display_sizes.session_height = description->session_height;

